### PR TITLE
Add hooks API docs

### PR DESCRIPTION
**Pull request by timleslie**

### CODE REVIEW

File: docs-next/pages/apis/hooks.mdx [hunk 1]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
 
 # Hooks API
 
-(coming soon)
+Hooks allow you to execute code at different stages of the mutation lifecycle when performing create, update, and delete operations.
+Lists and fields both support the same set of hook functions, with some slight differences in the arguments they accept.
+The differences will be explicitly called out below.
+For each hook, the fields hooks are applied to all fields in parallel first, followed by the list hooks.
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:8]: This line is super important. Can we format it differently so it gets called out more clearly?

e.g

```suggestion

!> For each hook, the fields hooks are applied to **all fields first** in parallel, followed by the list hooks.
```

tbh a diagram on this page would go a long way, I guess that's more for the guide though, and we can come back to that another time.

File: docs-next/pages/apis/hooks.mdx [hunk 2]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+
+### validateInput
+
+The `validateInput` function is used to validate the [`resolvedData`](#resolved-data-stages) that will be saved during a `create` or `update` operation.
+
+It is invoked after the `resolveInput` hooks have been run.
+
+If the `resolvedData` is invalid then the function should report validation errors with `addValidationError(msg)` for list hooks, or `addFieldValidationError(msg)` for field hooks.
+These error messages will be returned as a `ValidationFailureError` from the GraphQL API.
+
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:128]: Should we clarify (and is it correct) that if any validation errors are on the stack after this part of the lifecycle, the API will return without calling further lifecycle methods or performing any changes to the database?

File: docs-next/pages/apis/hooks.mdx [hunk 3]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+| `operation`     | The operation being performed (`'create'` or `'update'`).                                                                                                                         |
+| `originalInput` | The value of `data` passed into the mutation.                                                                                                                                     |
+| `existingItem`  | The currently stored item (`undefined` for `create` operations). This object is an unresolved list item. [list item API](./list-items) for more details on unresolved list items. |
+| `resolvedData`  | A [`resolved data`](#resolved-data-stages) object. The resolved data value after default values, relationship resolvers, and field resolvers have been applied.                   |
+| `context`       | The [`KeystoneContext`](./context) object of the originating GraphQL operation.                                                                                                   |
+
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:76]: General comment, I think we should explicitly define (or explain) the return type for each hook function as well as the arguments

File: docs-next/pages/apis/hooks.mdx [hunk 4]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+              listKey,
+              fieldPath,
+              operation,
+              originalInput,
+              existingItem,
+              resolvedData,
+              context,
+            }) => {
+              /* ... */
+              return resolvedData[fieldName];
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:110]: I'm unclear here as to what happens when the field input isn't 1:1 with the field path..?

@mitchellhamilton has a proposal on this, to make fields and input 1:1, which I think helps clarify this in the best way, but that's a thing for later

File: docs-next/pages/apis/hooks.mdx [hunk 5]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
+      hooks: {
+        validateInput: async ({
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:85]:
```suggestion
        resolveInput: async ({
```

File: docs-next/pages/apis/hooks.mdx [hunk 6]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+          context,
+        }) => {
+          /* ... */
+          return resolvedData;
+        },
+      },
+      fields: {
+        fieldName: text({
+          hooks: {
+            validateInput: async ({
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:100]:
```suggestion
            resolveInput: async ({
```

File: docs-next/pages/apis/hooks.mdx [hunk 7]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+
+Create and update operations take a `data` value for a single item from the GraphQL input and then perform a number of _data resolving_ steps before writing the final value to the database.
+
+At each stage of the data resolving process, the value of `resolvedData` can be modifed or augmented.
+The final value of `resolvedData` is the value that will be validated and saved to the database.
+
+The data resolving steps are applied in the following order:
+
+1. Initialisation: Set the value of `resolvedData` to the `data` input value from the GraphQL mutation.
+2. Defaults (built in): Any fields which have a default value and are `undefined` in `resolvedData` will be set to their default value.
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:447]: I believe this is only relevant for `create` operations? above, this section is scoped to "Create and update operations" so lets call this exception out if I'm right

File: docs-next/pages/apis/hooks.mdx [hunk 8]
```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
...
+3. Relationships (built in): The values for relationship fields on `resolvedData` are either an ID string (for `one` relationships), or an array of ID strings (for `many` relationships).
+   This is the format expected when saving relationship fields to the data.
```
Comment (JedWatson) [inline docs-next/pages/apis/hooks.mdx:449]: Did you mean "to the database"?

```suggestion
   This is the format expected when saving relationship fields to the database.
```

**Review by JedWatson — CHANGES_REQUESTED** [review]

File: docs-next/pages/apis/hooks.mdx (hunk 2 above)
Comment (timleslie) [reply docs-next/pages/apis/hooks.mdx:128]: Hmm, I would have thought that was obvious, but I can add a clarifying statement if you like.

File: docs-next/pages/apis/hooks.mdx (hunk 4 above)
Comment (timleslie) [reply docs-next/pages/apis/hooks.mdx:110]: Which part is unclear? Could you give an example of an input and which arguments could lead to confusion?

Comment (mitchellhamilton) [reply docs-next/pages/apis/hooks.mdx:113]: > I'm unclear here as to what happens when the field input isn't 1:1 with the field path..?

That doesn't/sort of can't happen for create/update.

**Review by JedWatson — APPROVED** [review]

**Pull request merged**